enh(blog): Add blog post on generative AI peer review policy (#734)
pyOpenSci/pyopensci.github.io: _posts/2025-09-16-generative-ai-peer-review.md

---
layout: single
title: "Navigating LLMs in Open Source: pyOpenSci's New Peer Review Policy"

> excerpt: "Generative AI tools are making is easier to generate large amounts of code which in some cases is causing a strain  on volunteer peer review programs like ours. Learn about pyOpenSci's policy on generative AI in peer review in this blog post."

Reviewer: ```suggestion
excerpt: "Generative AI products are reducing the effort and skill necessary to generate large amounts of code, which in some cases is causing a strain  on volunteer peer review programs like ours. Learn about pyOpenSci's policy on generative AI in peer review in this blog post."
```
Calling it a "tool" endorses some fitness for purpose, which is debatable.

author: "pyopensci"
permalink: /blog/generative-ai-peer-review-policy.html
header:
  overlay_image: images/headers/pyopensci-floral.png
categories:
  - blog-post
  - community
classes: wide
toc: true
comments: true
last_modified: 2025-09-16
---

authors: Leah Wasser, Mandy Moore, 

## Generative AI meets scientific open source

> It has been suggested that for some developers, using AI tools for tasks can increase efficiency by as much as 55%. But in open source scientific software, speed isn't everything—transparency, quality, and community trust matter just as much. So do the ethical questions these tools raise.

Reviewer: I wouldn't air such conjecture without citation (and believing in the integrity of that work). It could say that studies and perception are mixed and perhaps that perception of efficacy appears to exceed reality (citing the METR study).

**Edit this.** Whatever breakout content we want here.... needs to be all on a single line. 
{: .notice--success}

## Why we need guidelines

At [pyOpenSci](https://www.pyopensci.org/), we’ve drafted a new policy for our peer review process to set clear expectations around disclosing use of LLMs in scientific software packages.

This is not about banning AI tools. We recognize their value to some. Instead, our goal is transparency. We want maintainers to **disclose when and how they’ve used LLMs** so editors and reviewers can fairly and efficiently evaluate submissions.

## Our Approach: Transparency and Disclosure

> We know that people will continue to use LLMs. We also know they can meaningfully increase productivity and lower barriers to contribution for some. We also know that there are significant ethical, societal and other challenges that come with the development and use of LLM’s. 

Reviewer: Conjectures about the future depend greatly on legal outcomes and how society processes this moment. I would not say it's inevitable, but perhaps that pyOpenSci's policy will not on its own change the behavior of the community, especially those who aren't thinking about pyOpenSci.

Reviewer: ```suggestion
We acknowledge that social and ethical norms and concern for environmental and societal externalities varies greatly across the community, and yet few members of the community will look to pyOpenSci for guidance on whether to use LLMs in their own work. Our focus thus centers on assisting with informed decision-making and consent with respect to LLM use in the submission, reviewing, and editorial process.
```

I don't know if it's really true that pyOpenSci does not have the capability to influence decisions, in which case it could be worded that pyOpenSci advisors recommend against taking a position about whether LLM-generated content can be allowed in submissions. Or even a stronger stance that although pyOpenSci finds many uses of LLMs to run contrary to our community values, we believe that attempting to enforce a ban would incentivize dishonestly and erode community trust so severely that it outweighs epistemic benefits of such a policy, and therefore we seek only informed consent.

Our community’s expectation is simple: **be open about it**.

* Disclose LLM use in your README and at the top of relevant modules.  
* Describe how the tools were used   
* Be clear about what human review you performed.

Transparency helps reviewers understand context, trace decisions, and focus their time where it matters most.

### Human oversight

LLM-assisted code must be **reviewed, edited, and tested by humans** before submission.

* Run tests and confirm correctness.  
* Check for security and quality issues.  

> * Ensure style, readability, and clear docstrings.  

Reviewer 2: Ensure style, readability, clear, and concise docstrings.

Depending on the AI tool, generated docstrings can sometimes be overly verbose without adding meaningful understanding.

* Explain your review process in your software submission to pyOpenSci.

Please don’t offload vetting to volunteer reviewers. Arrive with human-reviewed code that you understand, have tested, and can maintain.

### Licensing awareness

> LLMs may be trained on mixed-license corpora. Outputs can create **license compatibility questions**, especially when your package uses a permissive license (MIT/BSD-3).

Reviewer: LLM output does not comply with the license of the input package, even when the input is permissively licensed (MIT, CC-BY), because it fails to comply with the attribution requirement of the license. The license of the package incorporating LLM output does not matter.

License compatibility only matters after an egregious violation is discovered: if the licenses are compatible, one could become compliant merely by adding attribution.

> LLMs may be trained on mixed-license corpora. Outputs can create **license compatibility questions**, especially when your package uses a permissive license (MIT/BSD-3).

Reviewer: MIT also requires attribution (emphasis added)
> The **above copyright notice** and this permission notice shall be included in all copies or substantial portions of the Software.

The point I was trying to make was that even if all sources for a given output have the same license, it's still a license violation because the copyright notice (attribution) is not preserved. Here's a crack at a more comprehensive description of the situation and possible remedies.

```suggestion
LLMs are trained on source code and documents with many licenses, most of which require attribution/preservation of a copyright notice (possibly in addition to other terms). LLM outputs sometimes produce verbatim or near-verbatim copies of [code](https://githubcopilotlitigation.com/case-updates.html) or [prose](https://arxiv.org/abs/2505.12546) from the training data, but with attribution stripped. Without attribution, such instances constitute a derivative work that violates the license, thus are likely to be copyright infringement and are certainly plagiarism. Copyright infringement and plagiarism are issues of process, not merely of the final artifact, so it is difficult to prescribe a reliable procedure for due diligence when working with LLM output, short of assuming that such output is always tainted and thus the generated code or derivative works can never come into the code base. We recognize that many users of LLM products for software development would consider such diligence impractical.

If similarities with existing software is detected **and** the licenses are compatible, one can come into compliance with the license by complying with its terms, such as by adding attribution. When the source package has an [incompatible license](https://dwheeler.com/essays/floss-license-slide.html), there is no simple fix. For example, if LGPL-2.1 code is emitted by an LLM into an Apache-2.0 project, no amount of attribution or license changes can bring the project into compliance. The Apache-2.0 project cannot even relicense to LGPL-2.1 without consent from every contributor (or their copyright holder). In such cases, the project would be responsible for deleting all implicated code and derivative works, and rewriting it all using [clean-room techniques](https://en.wikipedia.org/wiki/Clean-room_design).
```

* Acknowledge potential license ambiguity in your disclosure.  

> * Avoid pasting verbatim outputs that resemble known copyrighted code.  

Reviewer: How would someone determine this? Due diligence is to never use the output of an LLM directly, but that isn't how LLM-based coding products are marketed or used.

Reviewer: I honestly don't know either. In my opinion, a minimum of due diligence would be that any time a multi-line completion is accepted (and before editing it, since editing copied code would mean it is a derivative work), one runs a code search to see if there are any public results that are similar to the output. This is incredibly unrealistic because the interfaces of all coding LLM products are built to promote "don't know, don't care".

But to endorse the premise that "dunno; the LLM generated that" is a watertight defense with no ethical implications is to say that there is no ethical transgression other than not sufficiently covering one's tracks.

* Prefer human-edited, transformative outputs you fully understand.

We can’t control upstream model training data, but we can be cautious, explicit and critical about our usage.

### Ethics and inclusion

LLM outputs can reflect and amplify bias in training data. In documentation and tutorials, that bias can harm the very communities we want to support.

* Review AI-generated text for stereotypes or exclusionary language.  
* Prefer plain, inclusive language.  
* Invite feedback and review from diverse contributors.

Inclusion is part of quality. Treat AI-generated text with the same care as code.

## Supporting volunteer peer review

Peer review runs on **volunteer time**. Rapid, AI-assisted submissions can overwhelm reviewers—especially when code hasn’t been vetted.

* Submit smaller PRs with clear scopes.  
* Summarize changes and provide test evidence.  
* Flag AI-assisted sections so reviewers know where to look closely.  
* Be responsive to feedback, especially on AI-generated code.

These safeguards protect human capacity so high-quality packages can move through review efficiently.

## Benefits and opportunities

> LLMs are already helping developers:

Reviewer: ```suggestion
LLMs are already perceived as helping developers:
```
many maintainers/developers would claim they are misleading and cause more harm than good, even for these tasks

* Explaining complex codebases  
* Generating unit tests and docstrings  
* In some cases, simplifying language barriers for participants in open source around the world  
* Speeding up everyday workflows

> For some contributors, these tools make open source more accessible.

Reviewer: ```suggestion
Some contributors perceive these products as making open source more accessible.
```

> ## Challenges we must address

Reviewer 3: I know you mention it above in the human oversight section, but maybe it's important to add another section here explaining that LLMs frequently incorrectly do programming tasks (especially those that are slightly more complex!).

### Overloaded peer review

Peer review relies on volunteers. LLMs can produce large volumes of code quickly, increasing submissions with content that may not have been carefully reviewed by a human before reaching our review system.

### Ethical and legal complexities

> LLMs are often trained on copyrighted or licensed material. Outputs may create conflicts when used in projects under different licenses. They can also reflect extractive practices, like data colonialism, and disproportionately harm underserved communities.

Reviewer: ```suggestion
LLMs are often trained on copyrighted material with varying (or no) licenses. Outputs may constitute copyright infringement and/or ethical violations such as plagiarism. They can also reflect extractive practices, like data colonialism, and disproportionately harm underserved communities.
```
The licenses do not need to be different to be a license violation (and copyright infringement and/or plagiarism).

### Bias and equity concerns

AI-generated text can perpetuate bias. When it appears in documentation or tutorials, it can alienate the very groups open source most needs to welcome.

### Environmental impacts

> Training and running LLMs [requires massive energy consumption](https://www.technologyreview.com/2019/06/06/239031/training-a-single-ai-model-can-emit-as-much-carbon-as-five-cars-in-their-lifetimes/), raising sustainability concerns that sit uncomfortably alongside much of the scientific research our community supports.

Reviewer: The numbers are way higher now, and training is only one component.

### Impact on learning

Heavy reliance on LLMs risks producing developers who can prompt, but not debug or maintain, code—undermining long-term project sustainability and growth.

## What you can do now 

* **Be transparent.** Disclose LLM use in your README and modules.  

> * **Be accountable.** Thoroughly review, test, and edit AI-assisted code.  

Reviewer: What does it mean to be "accountable"? What sort of lapses would constitute misconduct and what are the consequences? (E.g., a lawyer can lose their job and be disbarred when their use of LLMs undermine the integrity of the court.)

---
layout: single
title: "Navigating LLMs in Open Source: pyOpenSci's New Peer Review Policy"
excerpt: "Generative AI products are reducing the effort and skill necessary to generate large amounts of code, which in some cases is causing a strain  on volunteer peer review programs like ours. Learn about pyOpenSci's policy on generative AI in peer review in this blog post."
author: "pyopensci"
permalink: /blog/generative-ai-peer-review-policy.html
header:
  overlay_image: images/headers/pyopensci-floral.png
categories:
  - blog-post
  - community
classes: wide
toc: true
comments: true
last_modified: 2025-09-16
---

> authors: Leah Wasser, Mandy Moore, 

Author: ```suggestion
authors: Leah Wasser, Jed Brown, Carter Rhea, Ellie Abrahams
```
